### PR TITLE
feat: add init service and init command

### LINKED ISSUE #452: create generic init implementation
### What needs to get done

Generic InitCommand should be created, so all the crafts can use it, that allows:
- specifying `name` as an option and `project_dir` as a positional argument (the way `cargo` / `uv` works)
- use Jinja2 for templates
- support profiles e.g. by searching `templates` for subdirectories and listing them as options
- by default it will allow to be run in non-empty directory, as long as it creates only new files (no-overrides). This behavior may be overwritten  (opt-out), so charmcraft will be able to error-out in non-empty directory

### Why it needs to get done

Currently all the applications have a separate way of handling `init`.
We should create a generic version, that everyone can use.
It will enable create specific profiles in snapcraft, make our codebase smaller and it will be easier to maintain in the single place.

### LINKED ISSUE #508: Import and add profile support for the `init` command
### What needs to get done

* Import the `init` command from its private location
* Add support for `--profile`
    * Profiles should be registered like craft-parts plugins

### Why it needs to get done

To allow snapcraft to create a profile for testing with `snapcraft init --profile test`.

Design notes: https://docs.google.com/document/d/1lK0tb1eAapbuRoWipJsJ5HdQ27XZk9CJLWqdE52SjPU/edit#heading=h.qrrrlg3fcxw

### PR DESCRIPTION
**Pull request by mr-cal**
- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox`?

-----

Broken into a few commits:
1. Imports the `init` command from its private location with only changes needed to satisfy the linters
2. Adds an `InitService`, a bunch of tests, and fixes some bugs caught when adding tests
3. Adds profile support

Due to the size, this may be easier to review the entire thing rather than per-commit.

Fixes #452
Fixes #508
Creates #530 
(CRAFT-3542)

### CODE REVIEW

File: craft_application/commands/init.py
Comment (mr-cal) [inline craft_application/commands/init.py:75]: I think we should fix #530 first.  Otherwise I would have to drop `choices=self.profiles`, which is an important piece.

File: craft_application/commands/__init__.py
Comment (lengau) [inline craft_application/commands/__init__.py:1]: Update the copyright pls

File: craft_application/commands/other.py
Comment (lengau) [inline craft_application/commands/other.py:21]: In that case why not put `InitCommand` into this file?

File: craft_application/util/file.py
Comment (lengau) [inline craft_application/util/file.py:22]:
```suggestion
from stat import S_IRGRP, S_IROTH, S_IRUSR
```

File: craft_application/services/init.py
Comment (lengau) [inline craft_application/services/init.py:154]: Why not just always copy the mode?

```suggestion
                file.chmod((template_dir / template_name).stat().st_mode)
```

That way the template author can set the file exactly how they want. It's also a closer match to the behaviour of `shutil.copy2` for files that don't need rendering.

In fact if you want to make it even closer:


```suggestion
            shutil.copystat(template_dir / template_name, path)
```

File: craft_application/services/init.py
Comment (lengau) [inline craft_application/services/init.py:176]: Why not just

```suggestion
        emit.debug(f"Creating project directory {str(project_dir)!r}.")
        project_dir.mkdir(parents=True, exist_ok=True)
```

File: craft_application/commands/init.py
Comment (lengau) [inline craft_application/commands/init.py:141]: We should probably convert this into an absolute path:

```suggestion
            return pathlib.Path(parsed_args.project_dir).expanduser().resolve()
```

File: docs/reference/changelog.rst
Comment (lengau) [inline docs/reference/changelog.rst:7]:
```suggestion
X.Y.0 (2024-MMM-DD)
```

Not sure how you feel about this, but that indicates that this branch now contains new features (so we can no longer make hotfixes off main).

**Review by lengau — CHANGES_REQUESTED** [review]

File: craft_application/commands/__init__.py
Comment (mr-cal) [reply craft_application/commands/__init__.py:1]: Is this still needed?  (source: https://chat.canonical.com/canonical/pl/3h9z4rwtmtf5uxrhyypppzofyr)

File: craft_application/services/init.py
Comment (mr-cal) [reply craft_application/services/init.py:154]: This is a great simplification, especially since the non-template files are already using `shutil.copy2()`.

File: craft_application/commands/other.py
Comment (tigarmo) [reply craft_application/commands/other.py:21]: I'm personally ok with leaving `InitCommand` in its own file, to keep this `other.py`'s size under control

**Review by tigarmo — APPROVED** [review]
Looks good :) I'm approving assuming @lengau 's comments will be addressed

File: craft_application/commands/other.py
Comment (mr-cal) [reply craft_application/commands/other.py:21]: That, keeping the test files under control, and I also like how the command and its tests are detached and easy to find in the file tree.

File: craft_application/commands/init.py
Comment (lengau) [inline craft_application/commands/init.py:88]: This is here rather than the service because we need the profiles before starting services, right?

**Review by lengau — APPROVED** [review]
LGTM, thanks!

File: craft_application/commands/init.py
Comment (mr-cal) [reply craft_application/commands/init.py:88]: Correct

**Comment by mr-cal** [comment]
Waiting on #530 to land first.

File: craft_application/commands/init.py
Comment (tigarmo) [reply craft_application/commands/init.py:75]: @mr-cal #530 is now fixed

**Pull request merged**
